### ISSUE TITLE
Planning for deepchem 1.3

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#669]
Now that deepchem 1.2 is almost ready to push out (#668), it's a good time to start planning for things to put into deepchem 1.3. Here are a few features I'd like to see in deepchem 1.3

- PyTorch support in main DeepChem. There are now two interesting PyTorch models in contrib (graph-convs and MPNNs). Moving these into DeepChem proper would be useful.
- Convert one-shot models into TensorGraph. I believe these are the last major class of non-TensorGraph models still in DeepChem.
- Deprecate all non-TensorGraph TensorFlow models. Lots of new users get very confused since they can't save non-TensorGraph TF models.

Once #668 is closed, I'd welcome contributions for any of these points.

What other features would be good for 1.3?

**Comment by rbharath** [comment]
Thought of another 1.3 target:

- Adding PDDBind Protein-protein and Protein-nucleic acid collections to `dc.molnet`. This should be a relatively easy task for a new DeepChem developer.
 
Once these issues have had a little time to mature, I'll open individual issues for each desired new feature.

**Comment by rbharath** [comment]
More thoughts:

- Uncertainty estimates on DeepChem models are the most common practitioner feature request I run into. Adding uncertainty estimate support onto DeepChem models would let users estimate domains of applicability for our models. Here are a few potential technical leads
  * Dropout based uncertainty (https://arxiv.org/pdf/1506.02142.pdf). Probably a good first method, but has a number of drawbacks, since the models need to be evaluated many times to gauge uncertainty. Also, have to forcibly introduce dropout into networks to use this technique (even when it may not help). H/T @patrickhop 
  * Bayesian Deep Learning (https://arxiv.org/pdf/1703.04977.pdf, H/T @cxhernandez). Appears to just require a small tweak to loss functions.

Any PRs adding implementations from these papers would be very cool.

**Comment by patrickhop** [comment]
I have dropout based uncertainty working, but I still need to write some tests. 

I agree with pretty much all your ideas. Some suggestions I have are more robust tests and refactoring the examples. Those are both pretty low lying fruit.

**Comment by rbharath** [comment]
@patrickhop Great! Looking forward to your PR

I'd definitely like to see more robust tests and refactored examples. Could we come up with concrete TODOs that a developer can perform?

**Comment by patrickhop** [comment]
@rbharath Perhaps some tests centered around core functionality? Training, saving, inference, etc.

**Comment by rbharath** [comment]
Don't our current tests cover that already? If not, we should absolutely cover it. Feel free to open an issue on this with more details on tests we're missing

**Comment by patrickhop** [comment]
I think we have coverage wrt to whether or not the code runs without crashing, but I don't think we look at the predictions coming out. Personally, I would like to go as far as having calibration molecules to test trained models against, but I would understand if there was resistance against that.

**Comment by rbharath** [comment]
I'd be up to test trained models with calibration molecules. We probably can't run on Travis, but we can run on the Jenkins build.

I think in general, we should begin to start offering pretrained models (BBBP, Solubility, etc.) This would be a great 1.4 target perhaps.

**Comment by lilleswing** [comment]
I think longer running correctness tests on Jenkins are important.  With moleculenet in a more stable state now running it weekly on Jenkins is actually a possibility.

Also I think we should start thinking about how to ship pre-trained models for people in industry.  A plug-in-play toxicity model would help a lot for getting our foot in the door and DeepChem installed.  To do that though we have to have benchmark correctness numbers.

**Comment by lilleswing** [comment]
In combination with the pre-trained models notes/guidlines for best practices for online training of the models in the face of new data.

**Comment by rbharath** [comment]
I'm strongly in favor of shipping pretrained models. How about a new module `dc.pretrained`? I'd like us to have pretrained models for

- Blood brain barrier penetration
- Cell Membrane permeability
- hERG
-  And more...

Please suggest more pretrained models we can put together.

**Comment by LRParser** [comment]
@rbarath is there any single pretrained model that you think might be easiest to ship? Maybe something based on tox21? I would be happy to help with one

**Comment by patrickhop** [comment]
@rbharath @lilleswing Wrt to pretrained models, I think we'll have to be selective about what gets put out there. Sadly, if someone in industry uses a solubility model that was trained on delaney that fails (and it probably will), they are not going to blame the delaney dataset, rather they will blame deep learning and deepchem.

Another concern is the source of truth. It depends on the assay, but assays across companies need not match. Lipo and kinetic solubility are probably safe. Clearance, BBB, and permeability probably are not.

**Comment by rbharath** [comment]
I think I'm fine with people complaining about us. Loud complaints are the fastest way to see where we're terrible and need to improve :)

Facetiousness aside, I agree we should do our best here. As @patrickhop points out, we will be judged for what we put out, so we should make a best faith effort.

One way to temper expectations might be for us to estimate domain of validity; if Tanimoto score of query is far from train, we print a loud warning saying there are no guarantees.

**Comment by evanfeinberg** [comment]
Overall, how much are we interested in incorporating PyTorch-based models
into DC 1.3? This applies generally and to the various ideas expressed
above.

On Tue, Jul 18, 2017 at 7:51 PM, Bharath Ramsundar <notifications@github.com

**Comment by peastman** [comment]
> Overall, how much are we interested in incorporating PyTorch-based models into DC 1.3?

My personal take on it: we really should be using TensorGraph whenever it's reasonable to do so.  There's a lot of benefit to having a unified architecture that everything's based on.  That way it will be much easier to combine different types of models with each other, and when we make improvements to that base architecture, all the different models benefit.

**Comment by lilleswing** [comment]
My personal take on it: we should make more and more things come for "free" with TensorGraph, and add utilities to add the same value that pytorch adds.  

Say we want to have a RL model with atomic convolutions -- we would then need a pytorch version of atomic convolutions, and a pytorch implementation of A3C.  Both of which while non-trivial in theory have lots of small details that need to be hammered out over time and manual testing (that we did).  You can say the same thing about any molecule specialty layer that we implement in TensorGraph.  A beauty of neural networks is you can use whole networks as parts of others.  If we fragment out code base we loose a lot of composability of networks.

Why are people interested in pytorch?  I'm honestly curious.  I believe that the tensorflow community is strong enough and enough people are making investments that we can shore up the shortcomings in the short term, and the community will deliver in the long term.  I could be wrong and pytorch will overtake tensorflow as the easy to use deep learning framework, but I am personally not buying it as of right now.

tl;dr 
I personally wouldn't encourage pytorch -- but I'm not discouraging it either.  Where people want to use pytorch I want to talk to them and ask them why so I can make those things easier to do in TensorGraph.

**Comment by rbharath** [comment]
This is good feedback. I do like the idea of centering our engineering efforts on TensorGraph. I think our goal is to build a system that is capable as serving as the computational foundations of pharmaceutical and biotech companies, and that will take lots of engineering. We're not such a big development team that we can afford to split our powers.

At the same time, PyTorch is really winning hearts and minds with researchers. It's much easier to experiment with and build exotic neural networks. TensorFlow has really doubled down on engineering, which makes it great for production Google-scale work, but it's not as friendly for algorithmic innovation.

I think there's room for both robust engineering and research code to co-exist. One way to do this might be to create a new repo in the deepchem github organization dedicated to pytorch models. Another way might be for us to leave PyTorch in contrib/ for now. We can always revisit the issue for the 1.4 release or later if PyTorch looks to be pulling ahead.

**Comment by rbharath** [comment]
As a more general note, the discussion on here has been great. Let's keep tossing around ideas on this thread.

In the meanwhile though, please feel free to pick any of the non-pytorch items on this thread and get cracking with implementation!

**Comment by lilleswing** [comment]
@rbharath where are the one-shot models in the code-base?

**Comment by rbharath** [comment]
https://github.com/deepchem/deepchem/blob/master/deepchem/models/tf_new_models/graph_models.py has `SequentialSupportGraph` which is the parent class for the one-shot models.

The one-shot layers are in https://github.com/deepchem/deepchem/blob/master/deepchem/nn/layers.py. `ResiLSTMEmbedding` and friends.

The low-data examples live in https://github.com/deepchem/deepchem/tree/master/examples/low_data

I think we should start emptying out `tf_new_models` as we get closer to 1.3.

**Comment by evanfeinberg** [comment]
I echo Bharath in noting the insightful discussion here. This is one of
those situations where there are strong advantages to different directions
(a good problem to have).

 A key point to consider is, the more we hitch our wagons to TensorFlow,
the harder it will get over time to ever switch to PyTorch as an underlying
engine. As Bharath noted, a significant portion of the research community
has transitioned to PyTorch due to its pliability and ease of use.
TensorFlow has the benefit of being able to take advantage of industrial
resources (e.g., Google's internal systems), but PyTorch is also faster for
smaller scale hardware. Pharma/biotech is a big industry but doesn't have
nearly the compute resources that a Google would have, and for the
foreseeable future will not be devoting large resources for deep learning.
We might want to put some serious thought into pivoting energies to
PyTorch. I'm not saying it's the single correct way forward, but it
certainly merits careful consideration.

@cxhernandez any comments?

**Comment by peastman** [comment]
> a significant portion of the research community has transitioned to PyTorch

Do you have any usage stats on that?  I hear about people playing around with it, but whenever I look up someone's code for actually doing something, it's almost always based on TensorFlow.

Remember, there will *always* be a hot new toy that everyone is excited about.  Right now that's PyTorch.  A year from now it could easily be something else.  But if you're trying to build a robust code and maintain it over the long term, you need to pick a technology and stick with it.  And preferably it should be a "boring" technology, something that's as mature and widely used as possible.  The current PyTorch version is described as "an early-release beta".  That's not something we want to rely on right now.

**Comment by cxhernandez** [comment]
My personal opinion is that PyTorch is *the* best framework out there to-date. This is mostly due to its outstanding readability, which also makes it the easiest framework with which to prototype models. If one of the goals of this project is to encourage (especially novice-level) users to contribute code, I think this is something that should be very much considered, as using PyTorch would lower the barrier-to-entry for contributing.

I can say that from my own experience— trying to parse the GraphConv code in DeepChem to port it into PyTorch— it took me a long time to even understand the TF codebase and, in fact, ended up being easier to just refer to the literature.

> Remember, there will always be a hot new toy that everyone is excited about. 

This is true, but PyTorch, as the name suggests, is based on Torch which has been around since 2002 and is backed by Facebook. It's not going anywhere.

**Comment by rbharath** [comment]
I certainly agree PyTorch is more beginner friendly. The main question though is whether enough engineering has happened to make it production ready. I believe Facebook itself uses PyTorch for experimentation and Caffe2 for deployment.

A good way to derisk this might be to build more Pytorch functionality in contrib/. For example, before adopting PyTorch into mainline, I'd like to see:

- A saving mechanism for PyTorch models
- Implementations of some standard layers for PyTorch DeepChem models. This could include atomic convs or graph conv variants.
- Some basic RL code in PyTorch

**Comment by rbharath** [comment]
Plus, it would help if PyTorch hit 1.0 or a stable release. We had many DeepChem breakages due to TF upstream changes and I'd prefer to avoid similar issues from PyTorch.

**Comment by lilleswing** [comment]
I see part of what DeepChem is attempting to do having a lot in common with OpenAI's RL baselines.
https://blog.openai.com/openai-baselines-dqn/

Do we know people on their team that we can chat with as to why they made that decision?  Also with OpenAI baselines in Tensorflow we kinda get all of their RL baselines with an easier port.

**Comment by cxhernandez** [comment]
> The main question though is whether enough engineering has happened to make it production ready.

Yeah, I wouldn't jump to fully support it just yet (see https://github.com/pytorch/pytorch/issues/2155 or https://github.com/pytorch/pytorch/issues/786). But I think there is value in exploring how to make the TensorGraph API framework-agnostic so that algorithms can be written in either TF or PyTorch.

**Comment by rbharath** [comment]
I'm against framework agnostic code. With my experience of Keras backend, it leads to spaghetti code. If we're doing this, we should make PyTorch support entirely separate of TensorGraph.

**Comment by rbharath** [comment]
It's starting to feel like time for the DeepChem 1.3 version bump. There have been a lot of changes in the last two months. What changes do we want to make sure are in before the version bump?

CC @lilleswing @peastman

**Comment by peastman** [comment]
I don't think there's anything really essential.  I'm working on #866, but I wouldn't hold up 1.3 just for that.

**Comment by rbharath** [comment]
I think I'd like to see our build issues resolved and #874 merged in before 1.3, but think other PRs can probably go in 1.4 if needed.

**Comment by rbharath** [comment]
The Travis issues are (probably temporarily) resolved. I think we're ready to go ahead. Does anyone have a reason we shouldn't make the DeepChem 1.3 cut? If no outstanding concerns, will go ahead at the end of the week.

**Comment by lilleswing** [comment]
We should get this one in.
https://github.com/deepchem/deepchem/pull/885

After that we should cut.

**Comment by rbharath** [comment]
Just made the cut! https://github.com/deepchem/deepchem/releases/tag/1.3.0

We still need to update docker/conda though.

**Comment by lilleswing** [comment]
Docker/Conda updated

**Comment by rbharath** [comment]
@lilleswing Awesome, thanks

Going to go ahead and close this issue.
